core/src/main/java/org/jboss/modcluster/mcmp/impl/DefaultMCMPHandler.java

```diff
@@ -800,6 +800,15 @@ void setEstablished(boolean established) {
         private synchronized Socket getConnection() throws IOException {
             if (this.socket == null || this.socket.isClosed()) {
                 this.socket = this.socketFactory.createSocket();
+
+                // If a proxy has been configured with a hostname, as opposed to an IP address, attempt to resolve the address again.
+                // Note that java.net.InetAddress.toString avoids a reverse name lookup.
+                // Also note that JVM caches DNS name lookups.
+                // See MODCLUSTER-728.
+                if (!this.socketAddress.isUnresolved() && (this.socketAddress.getAddress().toString().indexOf('/') > 0)) {
```

Comment:
Looks like it'd still resolve it, but would still want to check the proxy state (this.getState() != State.OK) here before creating the new socketAddress?  If we know the proxy is still currently OK, we should still be able to reuse the same socketAddress and so can skip the DNS.

Reply:
Can you point me to in which situation is the proxy in OK state yet the socket is to the proxy is closed?

Reply:
The socket stays open through an HTTP keepalive state, which could still be closed by the proxy/server side at any point.  So the HTTP keepalive socket being closed does not inherently mean the proxy is a !OK state.  If the proxy server is httpd, it could close a keepalive socket for a number of reasons like the KeepaliveTimeout being reached, a graceful restart, reaping of some child processes, closing out keepalive connections according to the AsyncRequestWorkerFactor with mpm_event to name a few.  In such cases, the socket is closed, but modcluster does not inherently face any issue to move the proxy from an OK state and so only needs to open a new socket with the existing socketAddress.

Reply:
Actually isUnresolved() is true when the address can't be resolved, if you are using a cloud and probably a service to get a httpd on resolvable hostname, the name won't be resolved. For example create an httpd service to resolve("myproxy") then "myproxy" won't resolved until it is started and running healthy. Additional it doesn't resolve when the pod is restarting and myproxy is a single replica httpd.

Reply:
testing:
+++
[jfclere@localhost X509]$ java MODCLUSTER728 httpd
InetSocketAddress getHostString: httpd
InetSocketAddress isUnresolved:true
InetSocketAddress getAddress:null
InetSocketAddress getHostName():httpd
Exception in thread "main" java.net.UnknownHostException: httpd
+++
then:
+++
[jfclere@localhost X509]$ java MODCLUSTER728 httpd
InetSocketAddress getHostString: httpd
InetSocketAddress isUnresolved:false
InetSocketAddress getAddress:httpd/10.33.144.1
InetSocketAddress getHostName():httpd
+++

Reply:
@aogburn You are right, updated the logic.

@jfclere I think you are right - that logic doesn't fix the issue if the proxy address was unresolvable.